libs/sdk-core/src/greenlight/node_api.rs

```diff
@@ -441,6 +438,32 @@ impl Greenlight {
         Ok(on_chain_balance)
     }
 
+    async fn pending_onchain_balance(&self, funds: cln::ListfundsResponse) -> Result<u64> {
+        let pending_onchain_balance = funds.channels.iter().fold(0, |a, b| match b.state() {
```

Comment:
I will add this again, do we want to add the outputs where it's not confirmed or reserved ?

Reply:
When looking at the behaviour of cln the outputs list are employed when channel opening is taking place. On the other hand when channel closes are happening it the `channels` field which are populated. The match statement covers also the force close attempt, the `AwaitingUnilateral` is triggered on force close. After this is done the the funds will be accessible for the peer how didn't initiate the force close. Which means it will be showing up in the listfunds. 

So it is under my impression that the log above also includes the force closed channels and it's sweep.



Here are the state changes for the peer who had their channel force closed on them .

```
"state_changes": [
            {
               "timestamp": "2024-01-16T10:39:03.829Z",
               "old_state": "CHANNELD_AWAITING_LOCKIN",
               "new_state": "CHANNELD_NORMAL",
               "cause": "user",
               "message": "Lockin complete"
            },
            {
               "timestamp": "2024-01-16T10:41:32.330Z",
               "old_state": "CHANNELD_NORMAL",
               "new_state": "AWAITING_UNILATERAL",
               "cause": "protocol",
               "message": "channeld: received ERROR error channel 607e907fccd85848a2e0a8cf25e010bf67771879afc064b099abd4cb7bab39f0: Forcibly closed by `close` command timeout"
            },
            {
               "timestamp": "2024-01-16T10:44:01.368Z",
               "old_state": "AWAITING_UNILATERAL",
               "new_state": "FUNDING_SPEND_SEEN",
               "cause": "onchain",
               "message": "Onchain funding spend"
            },
            {
               "timestamp": "2024-01-16T10:44:01.376Z",
               "old_state": "FUNDING_SPEND_SEEN",
               "new_state": "ONCHAIN",
               "cause": "onchain",
               "message": "Onchain init reply"
            }
         ],
"status": [
            "CHANNELD_NORMAL:Received error channel 607e907fccd85848a2e0a8cf25e010bf67771879afc064b099abd4cb7bab39f0: Forcibly closed by `close` command timeout",
            "ONCHAIN:Tracking their unilateral close",
            "ONCHAIN:All outputs resolved: waiting 99 more blocks before forgetting channel"
         ],
```

The peer closing the channel
```
"state_changes": [
            {
               "timestamp": "2024-01-16T10:39:03.827Z",
               "old_state": "CHANNELD_AWAITING_LOCKIN",
               "new_state": "CHANNELD_NORMAL",
               "cause": "remote",
               "message": "Lockin complete"
            },
            {
               "timestamp": "2024-01-16T10:40:55.058Z",
               "old_state": "CHANNELD_NORMAL",
               "new_state": "CHANNELD_SHUTTING_DOWN",
               "cause": "user",
               "message": "User or plugin invoked close command"
            },
            {
               "timestamp": "2024-01-16T10:40:59.887Z",
               "old_state": "CHANNELD_SHUTTING_DOWN",
               "new_state": "AWAITING_UNILATERAL",
               "cause": "user",
               "message": "Forcibly closed by `close` command timeout"
            },
            {
               "timestamp": "2024-01-16T10:44:03.927Z",
               "old_state": "AWAITING_UNILATERAL",
               "new_state": "FUNDING_SPEND_SEEN",
               "cause": "user",
               "message": "Onchain funding spend"
            },
            {
               "timestamp": "2024-01-16T10:44:03.931Z",
               "old_state": "FUNDING_SPEND_SEEN",
               "new_state": "ONCHAIN",
               "cause": "user",
               "message": "Onchain init reply"
            }
         ],
         "status": [
            "ONCHAIN:Tracking our own unilateral close",
            "ONCHAIN:3 outputs unresolved: in 5 blocks will spend DELAYED_OUTPUT_TO_US (1b916ba0213d85138af0d67760d05aa26591975e2fc6d230ba7f21fb2e944fbb:0) using OUR_DELAYED_RETURN_TO_WALLET"
         ],
```